packages/react-router/modules/matchPath.js

```diff
@@ -26,11 +27,13 @@ const compilePath = (pattern, options) => {
 /**
  * Public API for matching a URL pathname to a path pattern.
  */
-const matchPath = (pathname, path, options = {}) => {
+const matchPath = (pathname, path, options = {}, parent = null) => {
   const { exact = false, strict = false } = options
 
   if (!path)
-    return { url: pathname, isExact: true, params: {} }
+    return { url: parent ? parent.url : '/', isExact: true, params: {} }
+
+  path = simpleResolve(path, parent && parent.url ? parent.url : '')
```

Comment:
Resolve using `parent.url` instead of `parent.path` for a simpler regexp comparison. Below, the parent's params are assigned to the `params` object so that the child can also reference them.

```js
<Route path=':category' render={() => (
  <Route path=':id' render={({ params }) => (
    console.log(params), // { category: 'lamps', id: 'i<3' }
    null
  )} />
)} />
```